core/src/handshake/manager/internal_engine.rs

```diff
@@ -153,15 +234,40 @@ fn match_orders(
         quote_amount,
         direction: order1.side.into(),
         max_minus_min_amount,
-        min_amount_order_index: if order1.amount <= order2.amount { 0 } else { 1 },
+        min_amount_order_index: if party0_max_amount <= party1_max_amount {
+            0
+        } else {
+            1
+        },
     }))
 }
 
+/// Compute the maximum matchable amount for an order and balance
+fn compute_max_amount(price: &FixedPoint, order: &Order, balance: &Balance) -> u64 {
```

Comment:
I thought about it, but that's a testing helper and therefore a much less stable interface to depend on. We also can't put it here and import in `circuits` because that would create a circular dependency. I think the long term best solution is to factor all of this out into crates, one of which may expose primitives like a plaintext match.